src/main/java/com/google/sps/servlets/RecommendServlet.java

```diff
@@ -0,0 +1,159 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.servlets;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.EntityNotFoundException;
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.datastore.KeyFactory;
+import com.google.appengine.api.datastore.Query;
+import com.google.gson.Gson;
+import com.google.sps.Firebase;
+import com.google.sps.Interactions;
+import com.google.sps.Recommend;
+import com.google.sps.Utils;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.TreeMap;
+import java.util.logging.Logger;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/recommend")
+public class RecommendServlet extends HttpServlet {
+
+  private static final Logger LOGGER = Logger.getLogger(RecommendServlet.class.getName());
+  private static final int EVENTS_LIMIT = 10;
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    final DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    final Gson gson = new Gson();
+    response.setContentType("application/json");
+
+    String userToken = request.getParameter("userToken");
+    if (userToken == null || !Firebase.isUserLoggedIn(userToken)) {
+      response.getWriter().println(gson.toJson(new ArrayList<Entity>()));
+      return;
+    }
+
+    List<Entity> events = new ArrayList<>();
+    String userID = Firebase.authenticateUser(userToken);
+    Key recKey = KeyFactory.createKey("Recommendation", userID);
+    Entity recommendations = null;
+    try {
+      // find previously computed recommendations
+      recommendations = datastore.get(recKey);
+      List<Long> recIds = (List<Long>) recommendations.getProperty("recs");
+      int limit = EVENTS_LIMIT;
+      for (int i = 0; i < recIds.size() && i < limit; i++) {
+        long eventId = recIds.get(i);
+        Key eventKey = KeyFactory.createKey("Event", eventId);
+        try {
+          events.add(datastore.get(eventKey));
+        } catch (EntityNotFoundException eventNotFound) {
+          LOGGER.info("event " + eventId + " not found.");
+          limit++;
+        }
+      }
+    } catch (EntityNotFoundException | NullPointerException exception) {
+      // no prev recommendations, compute simple recommendations based on tags and distance only
+      LOGGER.info("No recommendations found for " + userID + ". Computing now.");
+      recommendations = new Entity("Recommendation", userID);
```

Comment:
It's cleaner put the computation of recommendations into another method.

Reply:
Fixed!